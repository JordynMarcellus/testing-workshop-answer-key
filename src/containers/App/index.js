import React, { useState } from "react";
import { FormContainer } from "../FormContainer";
import { SinglePokemonContainer } from "../SinglePokemonContainer";
import * as AppStyles from "./index.styles";
import { Global, css } from "@emotion/core";

const GlobalStyles = css`
  html,
  body {
    background-color: #eb2f06;
    font-size: calc(14px + (26 - 14) * ((100vw - 300px) / (1920 - 300)));
  }
`;

const validateIsDisabled = inputValue => (inputValue === "" ? true : false);

const App = () => {
  const [searchInputValue, onSearchInputChange] = useState("");
  const [submittedValueState, onFormSubmit] = useState(null);
  const isDisabled = validateIsDisabled(searchInputValue);
  const handleChange = event => onSearchInputChange(event.target.value);
  const handleSubmit = event => {
    event.preventDefault();
    onFormSubmit(searchInputValue);
  };

  return (
    <AppStyles.Layout className="App">
      <Global styles={GlobalStyles} />
      <FormContainer
        {...{
          handleChange,
          handleSubmit,
          isDisabled,
          searchInputValue,
        }}
      />
      <AppStyles.CharacterDisplay>
        {submittedValueState ? (
          <SinglePokemonContainer pokemonName={submittedValueState} />
        ) : null}
      </AppStyles.CharacterDisplay>
    </AppStyles.Layout>
  );
};

export default App;
