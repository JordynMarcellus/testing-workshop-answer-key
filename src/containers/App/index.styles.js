import styled from "@emotion/styled";

export const Layout = styled.main`
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

export const CharacterDisplay = styled.section`
  background: white;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
`;
