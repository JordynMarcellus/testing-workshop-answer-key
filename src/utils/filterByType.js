//TODO: past jord note -- don't pass in an empty array in the final code. this is a subtle bug that will show why it's good to test ðŸ¤“
export const filterByType = (arrayOfPokemon, type) =>
  arrayOfPokemon.reduce((filteredPokemonArray, pokemonAtArrayIndex) => {
    if (pokemonAtArrayIndex.types.includes(type)) {
      return [...filteredPokemonArray, pokemonAtArrayIndex];
    }
    return filteredPokemonArray;
  });
