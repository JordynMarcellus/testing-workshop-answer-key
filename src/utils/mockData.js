export const mockPikaData = {
  data: {
    name: "pikachu",
    types: [
      {
        type: {
          name: "electric",
        },
      },
    ],
    sprites: {
      front_default: "path/to/img",
    },
  },
};

export const mockGolemData = {
  data: {
    name: "golem",
    types: [
      {
        type: {
          name: "rock",
        },
      },
      {
        type: {
          name: "ground",
        },
      },
    ],
    sprites: {
      front_default: "path/to/img",
    },
  },
};
