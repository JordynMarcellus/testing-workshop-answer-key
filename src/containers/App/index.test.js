import React from "react";
import App from "./";
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
} from "react-testing-library";
import axios from "axios";
import "jest-dom/extend-expect";
import { mockPikaData } from "../../utils/mockData";

jest.mock("axios");

describe("App integration tests ðŸ¤ª", () => {
  afterEach(() => {
    cleanup();
    axios.mockReset();
  });
  it("enters text into the form, clicks submits and shows a pokemon", async () => {
    const originalError = console.error;
    console.error = jest.fn();
    axios.mockResolvedValueOnce(mockPikaData);
    const { getByLabelText, getByText, container } = render(<App />);
    fireEvent.change(getByLabelText("Search for a single Pokemon"), {
      target: { value: "Pikachu" },
    });
    fireEvent.click(getByText("Search"));
    await waitForElement(() => getByText("pikachu"));
    expect(container).toHaveTextContent("pikachu");
    expect(container.firstChild).toMatchSnapshot();
    console.error = originalError;
  });
  it("enters text into the form, clicks submits and shows a pokemon", async () => {
    const originalError = console.error;
    console.error = jest.fn();
    axios.mockResolvedValueOnce(mockPikaData);
    const { getByLabelText, getByText } = render(<App />);
    fireEvent.change(getByLabelText("Search for a single Pokemon"), {
      target: { value: "Pikachu" },
    });
    fireEvent.click(getByText("Search"));
    await waitForElement(() => getByText("pikachu"));
    fireEvent.change(getByLabelText("Search for a single Pokemon"), {
      target: { value: "Pika" },
    });
    expect.assertions(1);
    expect(axios).toHaveBeenCalledTimes(1);
    console.error = originalError;
  });
});
