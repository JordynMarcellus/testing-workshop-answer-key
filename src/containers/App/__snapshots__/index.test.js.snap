// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App integration tests ðŸ¤ª enters text into the form, clicks submits and shows a pokemon 1`] = `
.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #f1f1f1;
  padding: 2rem;
}

.emotion-0 {
  padding: 0.25rem;
  border: 2px solid #333333;
  border-radius: 5px;
  margin: 10px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: hidden;
  margin: 10px;
  padding: 12px 12px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all 150ms linear;
  transition: all 150ms linear;
  text-align: center;
  white-space: nowrap;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  text-transform: none;
  text-transform: capitalize;
  color: #333;
  background-color: #dfe6e9;
  border: 0 none;
  border-radius: 4px;
  font-size: 13px;
  font-weight: 500;
  line-height: 1.3;
  -webkit-appearance: none;
  -moz-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 0 0 160px;
  -ms-flex: 0 0 160px;
  flex: 0 0 160px;
  box-shadow: 2px 5px 10px var(--color-smoke);
}

.emotion-2:hover {
  -webkit-transition: all 150ms linear;
  transition: all 150ms linear;
  opacity: 0.85;
}

.emotion-2:active {
  -webkit-transition: all 150ms linear;
  transition: all 150ms linear;
  opacity: 0.75;
}

.emotion-2:focus {
  outline: 1px dotted #959595;
  outline-offset: -4px;
}

.emotion-10 {
  background: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-width: 300px;
}

<main
  class="App emotion-12 emotion-13"
>
  <form
    class="emotion-6 emotion-7"
  >
    <h1>
      PokÃ©-picker
    </h1>
    <label
      for="search-input"
    >
      Search for a single Pokemon
    </label>
    <input
      class="emotion-0 emotion-1"
      id="search-input"
      value="Pikachu"
    />
    <div
      class="emotion-4 emotion-5"
    >
      <button
        class="emotion-2 emotion-3"
        data-testid="single-search-button"
        id="single-search-button"
      >
        Search
      </button>
    </div>
  </form>
  <section
    class="emotion-10 emotion-11"
  >
    <article
      class="emotion-8 emotion-9"
    >
      <h1>
        pikachu
      </h1>
      <ul>
        <span>
          electric
        </span>
      </ul>
      <img
        alt="My chonky fave pikachu"
        src="path/to/img"
        title="My chonky fave pikachu"
      />
    </article>
  </section>
</main>
`;
